Google.Api.Gax.Grpc/RetrySettings.cs

```diff
@@ -179,6 +181,24 @@ public TimeSpan NextBackoff(TimeSpan currentBackoff)
             }
         }
 
+        /// <summary>
+        /// Returns a sequence of already-jittered backoffs. A caller can just iterate over this
+        /// sequence and delay by each returned timespan. This sequence is infinite; callers do not
```

Comment:
Couldn't the sequence respect MaxAttempts value and be finite? If done with a foreach as Chris suggested, the calling code would be simpler I think.

Reply:
Unfortunately that's pretty hard - because when we fall off the end, we really want to just not catch the exception. (We don't want to catch it and then rethrow it later, losing a bunch of context, IMO.)
I do see the appeal, but the nature of exceptions is working against us :(

Reply:
True, to keep context we would still have to keep track of the attempts on the calling code. And I do agree that context is more important here.

Reply:
I'm now starting to think that there's so little in the loop, we can probably put an "execute with retry" call into RetrySettings after all. The situation is simpler than it was. Hmm.

Reply:
So far, we've had an informal agreement to have no extra public methods in "settings" classes - they are pure data. This would break that. Not that we can't do it, but we should consciously do it if we choose to.

To continue the foreach/max-attempts idea:
If the foreach loop  looped over `retrySettings.GetJitteredBackoffSequence().Append(TimeSpan.Zero)`, then the `catch when` condition can include checking that, and it's all nice and elegant :) (In my opinion anyway!)

Reply:
Will keep noodling on this. I suspect there's stuff we can do here :)

Reply:
What if the the sequence is not of TimeSpan but of RetryAttempt (or whatever other name that's better)? And
```
IEnumerable<RetryAttempt> GetJitteredBackoffSequence(Expiration overallDeadline, IScheduller, IClock);
class RetryAttempt
{    
    bool CanRetry(RpcException e) {// MaxAttempts, Filter, OverallDeadline all taken into account} // 
    Task Delay(); // Throws if above was false when fisrt called (because of OverallDeadline)
}

```
The last element of the sequence is the first one that returns false on `CanRetry`.

If we could `catch ... when await ...` that'd be great, then we could have only one method Task<bool> TryDelay(RpcException) :)

Reply:
Will probably come back to all of this next week. I like that idea, but want to get other stuff done first :)